[Fix] specify 'origin' remote name with `git clone ...` installs

rivy opened the pull request:
> I ran into this as an installation failure when installing `nvm` on a PC using some `git` configuration customizations.

Since `git` may be configured locally to use a non-'origin' default remote name, specify 'origin' as the remote name when cloning to get the expected local repo setup.

ljharb commented:
Thanks, this is a great catch!

Unfortunately, i believe git didn't add this option until v2.29, and we currently only require git v1.7.10. Would there be an alternative approach where instead of hardcoding "origin", we derive the default remote name, and just use that throughout the install script?

rivy commented:
Hmm.

To my knowledge, there is no real default remote after the repo is created. I don't believe `git` holds any special regard for 'origin' as a default (other than then initial remote name). If there are more than one, `git remote` just displays all of them in alpha order.

A couple of ways to deal with the issue that come to mind are just using the `git ...` command as amended in the PR followed by the original version as an alternate. From looking at the docs, I believe that prior to 1.7.10, there was no way to clone a repo to anything other than the 'origin' remote. And the `--origin ...` option was added in that version along with the configuration variable 'clone.defaultRemoteName'. So, if `command git clone "$(nvm_source)" --depth=1 --origin=origin "${INSTALL_DIR}"` fails then `command git clone "$(nvm_source)" --depth=1 "${INSTALL_DIR}"` should always install the remote as 'origin'.

Then using, `command git clone "$(nvm_source)" --depth=1 --origin=origin "${INSTALL_DIR}" || command git clone "$(nvm_source)" --depth=1 "${INSTALL_DIR}" || { ... }` should always set up the repo to have the remote 'origin' pointing at the expected URL.

Alternatively, the script could `cd "${INSTALL_DIR" ; git remote rename $(git remote) origin` immediately after the clone (without  a `--origin=origin` flag). That should also work for any version and set up the remote 'origin' correctly. (`git` doesn't throw an error if you rename a remote to the same name.)

I think the first option is likely the better choice. It should work for any version, be localized to a single code location, and should always result in the expected 'origin' setup.

rivy commented:
I pushed a modified version.

ljharb commented:
git docs say `Overrides clone.defaultRemoteName from the config.` which suggests that `git config clone.defaultRemoteName || echo origin` would be the value to use?

rivy commented:
> git docs say `Overrides clone.defaultRemoteName from the config.` which suggests that `git config clone.defaultRemoteName || echo origin` would be the value to use?

Yep, but that's only really good at the time of cloning.

ljharb commented:
ah, so you're saying that that would work for the initial clone, but wouldn't help with future git operations?

wouldn't we be able to assume that the default, if configured, is unlikely to change?

rivy commented:
I added references and corrected the mentioned versions to my second post. And, in looking at them again, it looks like `-o` was added in v1.1.0 [ref: <https://github.com/git/git/commit/e6c310fd0d7384973efc6b1d5999a5e8a5b2f3bd>], though I don't see the where `--origin` was added.

I could just simply change it to use `-o origin` and that would, I believe, fit the version restriction.

rivy commented:
> ah, so you're saying that that would work for the initial clone, but wouldn't help with future git operations?
> 
> wouldn't we be able to assume that the default, if configured, is unlikely to change?

If some user changed it, I think it's actually *more likely* to change in the future. And the change will be external and unknown to the repo. I think that setting it up with the expected 'origin' name from the start is going to be much more robust. Much less likely that a user would go into the repo and rename it than rethinking the config choice.

Personally, I had 'main' for a while, then went with 'local' on my home PC because I have a lot of customized soft-forks.

But, I'm likely a 0.1%-er here.

ljharb commented:
alrighty, let's go with `-o origin`, and just hardcode it, then.

ljharb commented on install.sh:
```diff
@@ -163,7 +163,8 @@ install_nvm_from_git() {
       # Cloning repo
-      command git clone "$(nvm_source)" --depth=1 "${INSTALL_DIR}" || {
+      command git clone "$(nvm_source)" --depth=1 -o origin "${INSTALL_DIR}" 2> /dev/null \
+        || command git clone "$(nvm_source)" --depth=1 "${INSTALL_DIR}" || {
```
```suggestion
        || command git clone "$(nvm_source)" --depth=1 -o origin "${INSTALL_DIR}" || {
```

rivy replied:
I'll just shrink it to the single command/line since at least `-o origin` looks like it'll work with the version restriction.

ljharb replied:
The difference is that the first one suppresses stderr; it's intentional to have the command twice so that when it fails, it tries again and prints out stderr for debugging purposes.

mshtaqalshybany-maker replied:
OKX

ljharb reviewed (commented):
Can there be any install tests that cover this?

rivy commented:
> Can there be any install tests that cover this?

Sure. I'll take a look at your tests and try to work up a something that fits.

I think I should be able to add something to [install_nvm_from_git](https://github.com/nvm-sh/nvm/blob/2b5e53fcd57409e018b468fd05eebad06fbff3fa/test/install_script/install_nvm_from_git). 

Are you testing with the minimum git version? I don't see that anywhere.

ljharb commented:
No, git version is just an implicit requirement.

vincent89291 reviewed (commented):
Explicit
